src/contrasts.jl

```diff
@@ -0,0 +1,215 @@
+
+# Constructors
+
+```julia
+ContrastsMatrix(contrasts::AbstractContrasts, levels::AbstractVector)
+ContrastsMatrix(contrasts_matrix::ContrastsMatrix, levels::AbstractVector)
+```
+
+# Arguments
+
+* `contrasts::AbstractContrasts`: The contrast coding system to use.
+   We only use `DummyCoding()`
+
+* `levels::AbstractVector`: The levels to generate contrasts for.
+
+* `contrasts_matrix::ContrastsMatrix`: Constructing a `ContrastsMatrix` from
+  another will check that the levels match.  This is used, for example, in
+  constructing a model matrix from a `ModelFrame` using different data.
+"""
+function ContrastsMatrix(
+    contrasts::C, levels::AbstractVector{T}
+) where {C<:AbstractContrasts,T}
+
+    # if levels are defined on contrasts, use those, validating that they line up.
+    # what does that mean? either:
+    #
+    # 1. DataAPI.levels(contrasts) == levels (best case)
+    # 2. data levels missing from contrast: would generate empty/undefined rows.
+    #    better to filter data frame first
+    # 3. contrast levels missing from data: would have empty columns, generate a
+    #    rank-deficient model matrix.
+    c_levels = something(DataAPI.levels(contrasts), levels)
+    if eltype(c_levels) != eltype(levels)
+        throw(
+            ArgumentError(
+                "mismatching levels types: got $(eltype(levels)), expected " *
+                "$(eltype(c_levels)) based on contrasts levels.",
+            ),
+        )
+    end
+    mismatched_levels = symdiff(c_levels, levels)
+    if !isempty(mismatched_levels)
+        throw(
+            ArgumentError(
+                "contrasts levels not found in data or vice-versa: " *
+                "$mismatched_levels." *
+                "\n  Data levels: $levels." *
+                "\n  Contrast levels: $c_levels",
+            ),
+        )
+    end
+
+    n = length(c_levels)
+    if n == 0
+        throw(
+            ArgumentError(
+                "empty set of levels found (need at least two to compute " * "contrasts)."
+            ),
+        )
+    elseif n == 1
+        throw(
+            ArgumentError(
+                "only one level found: $(c_levels[1]) (need at least two to " *
+                "compute contrasts).",
+            ),
+        )
+    end
+
+    # find index of base level. use contrasts.base, then default (1).
+    base_level = baselevel(contrasts)
+    baseind = base_level === nothing ? 1 : findfirst(isequal(base_level), c_levels)
+    if baseind === nothing
+        throw(ArgumentError("base level $(base_level) not found in levels " * "$c_levels."))
+    end
+
+    tnames = termnames(contrasts, c_levels, baseind)
+
+    mat = contrasts_matrix(contrasts, baseind, n)
+
+    return ContrastsMatrix(mat, tnames, c_levels, contrasts)
+end
+
+function ContrastsMatrix(c::Type{<:AbstractContrasts}, levels::AbstractVector)
+    return throw(
+        ArgumentError("contrast types must be instantiated (use $c() instead of $c)")
+    )
```

Comment:
```suggestion
    msg = "contrast types must be instantiated (use $c() instead of $c)")
    throw(ArgumentError(msg))
```
`return` statement was never reached I think